File: history-service/models/history.go
package models

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
	"time"
)

type QueryHistory struct {
	ID           uuid.UUID `json:"id" gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	UserID       uuid.UUID `json:"user_id" gorm:"type:uuid;not null"`
	DatabaseUUID uuid.UUID `json:"database_uuid" gorm:"type:uuid;not null"`
	QueryType    string    `json:"query_type" gorm:"type:varchar(50);not null"`
	Query        string    `json:"query" gorm:"type:text;not null"`
	Result       string    `json:"result" gorm:"type:jsonb"`
	Timestamp    time.Time `json:"timestamp" gorm:"autoCreateTime"`
	Success      bool      `json:"success" gorm:"not null;default:false"`
}

func (q *QueryHistory) BeforeCreate(tx *gorm.DB) (err error) {
	q.ID = uuid.New()
	return
}

File: diploma-frontend/src/lib/api/history.ts
import api from './index';

export async function getHistory(queryString: string = ''): Promise<PaginatedResponse> {
    const url = queryString ? `/history?${queryString}` : '/history';
    return await api.get(url);
}

export const clearHistory = async (): Promise<boolean> => {
    const response = await api.delete('/history');
    return response.status === 200;
};

export interface QueryHistory {
    id: string;
    user_id: string;
    database_uuid: string;
    query_type: 'SQL' | 'Natural Language';
    query: string;
    result: any;
    timestamp: string;
    success: boolean;
}

export interface PaginatedResponse {
    data: QueryHistory[];
    page: number;
    per_page: number;
    total: number;
    last_page: number;
}

export async function createHistoryEntry(entry: {
    endpoint: string;
    timestamp: string;
    query: any;
    result: any;
    success: boolean;
}) {

    const payload = {
        query_type: entry.endpoint.includes('text-to-sql')
            ? 'text-to-sql'
            : entry.endpoint.includes('execute-sql')
                ? 'execute-sql'
                : 'database',
        query: typeof entry.query === 'string' ? entry.query : JSON.stringify(entry.query),
        result: typeof entry.result === 'string' ? entry.result : JSON.stringify(entry.result),
        timestamp: entry.timestamp,
        success: entry.success
    };

    await api.post('/history', payload);
    console.log("Added to history")
}
